torchtitan/experiments/torchcomms/parallel_dims.py

```diff
@@ -34,25 +36,107 @@ def _build_mesh_without_ep(self) -> DeviceMesh:
         logger.info(f"Building {len(dims)}-D device mesh with {names}, {dims}")
         backend = os.environ["TEST_BACKEND"]
         device = torch.device("cuda")
-        # TODO:
-        # - Extend support for additional parallelism strategies (e.g., pipeline, context)
-        # - Refactor and modularize initialization logic for communication objects and device mesh construction.
-        if (
-            self.dp_shard > 1
-            and self.pp == 1
-            and self.dp_replicate == 1
-            and self.cp == 1
-            and self.tp == 1
-        ):
-            self.comms = []
-            comm = torchcomms.new_comm(backend, device, name="main")
-            # TODO: it's a hacky solution for now and we will update it in a week
-            mesh = init_device_mesh(
-                mesh_dim_comms=(comm, comm, comm, comm),
-                mesh_dim_names=("dp_shard", "dp", "dp_cp", "dp_shard_cp"),
+        mesh = torch.arange(self.world_size, dtype=torch.int, device="cpu").view(
+            self.pp, self.dp_replicate, self.dp_shard, self.cp, self.tp
+        )
+        comm = torchcomms.new_comm(
+            backend,
+            device,
+            name="comms_test_n_d_parallel",
+            timeout=datetime.timedelta(seconds=60),
+        )
+
+        # Get current rank to determine which groups this rank belongs to
+        cur_rank = comm.get_rank()
+
+        mesh_dim_names = ["pp", "dp_replicate", "dp_shard", "cp", "tp"]
+        ranks_per_dim = {}
+        comm_per_dim = {}
+        for idx, dim_name in enumerate(mesh_dim_names):
+            # Calculate global ranks mapping for this mesh dimension
+            global_ranks = mesh.transpose(idx, -1).reshape(-1, mesh.size(idx)).tolist()
+            for row in global_ranks:
+                if cur_rank in row:
+                    ranks_per_dim[dim_name] = row
+                    break
+
+        # Create communicators using the new single-list API
+        for dim_name, ranks in ranks_per_dim.items():
+            comm_per_dim[dim_name] = comm.split(ranks, dim_name)
+
+        try:
+            device_mesh = init_device_mesh(
+                mesh_dim_comms=(
+                    comm_per_dim["pp"],
+                    comm_per_dim["dp_replicate"],
+                    comm_per_dim["dp_shard"],
+                    comm_per_dim["cp"],
+                    comm_per_dim["tp"],
+                ),
+                mesh_dim_names=("pp", "dp_replicate", "dp_shard", "cp", "tp"),
                 _global_comm=comm,
             )
-            self.comms.append(comm)
-            return mesh
-        else:
-            raise NotImplementedError("Only support FSDP 1D parallelism for now.")
+        except TypeError as e:
+            # TODO: remove this once PT 2.10 is released
+            if "_rank" in str(e):
+                comm.finalize()
+                for sub_comm in comm_per_dim.values():
+                    sub_comm.finalize()
+                return
```

Comment:
I think the order is wrong. comm.finalize() needs to be after sub_comm.finalize(). This was fixed in trunk already.